pages/en/lb3/Discovering-models-from-relational-databases.md

```diff
@@ -23,69 +23,182 @@ This process is called _discovery_ and is supported by the following connectors:
-      console.log("\nProduct: ", prod);
-      console.log("\n ------------- ");
+{% include code-caption.html content="/server/boot/discover-models.js" %}
+```js
+'use strict';
+
+module.exports = function(app, callback) {
+  // Obtain the datasource registered with the name "db"
+  const dataSource = app.dataSources.db;
+
+  // Step 1: define a model for "salaries" table,
+  // including any models for related tables (e.g. "employees").
+  dataSource.discoverAndBuildModels(
+    'salaries',
+    {relations: true},
+    function(err, models) {
+      if (err) return callback(err);
+
+      // Step 2: expose all new models via REST API
+      for (const modelName in models) {
+        app.model(models[modelName], {dataSource: dataSource});
+      }
+
+      callback();
     });
-  });
-});
+};
+```
+
+Next steps:
+
+ 1. Start your application in the usual way, e.g. via `npm start` or `node .`
+ 2. Use [API Explorer](./Use-API-Explorer.md) to inspect your dynamically
+    created REST API.
+
+## Discover and save model definitions
+
+Use [dataSource.discoverSchema()](http://apidocs.loopback.io/loopback-datasource-juggler/#datasource-prototype-discoverschema)
+API to obtain a model definition for a given database table.
+
+To leverage this API, you need to write a small script to discover the models
+you are interested in and write them to model JSON files.
+
+Let's walk through an example implementation of such script.
+
+{% include note.html content="
+The code examples in this section are using features introduced in
+Node.js 8.x LTS. Users running on an older version of Node.js need
+to rework the code to use promises (or callbacks) instead of async/await
+for flow control.
+" %}
+
+First of all, we need to import APIs we will be using later:
+
+{% include code-caption.html content="/discover-models.js" %}
+```js
+'use strict';
+
+const loopback = require('loopback');
+const promisify = require('util').promisify;
+const fs = require('fs');
+const writeFile = promisify(fs.writeFile);
+const readFile = promisify(fs.readFile);
+const mkdirp = promisify(require('mkdirp'));
+```
+
+Secondly, we need to initialize the data source backed by the database
+we want to discover our models from. While it's possible to load the full
+application and let loopback-boot load all artifacts (including data sources),
+we recommend a simpler and faster approach of creating the data source
+instance explicitly:
+
+```js
+const DATASOURCE_NAME = 'db';
+const dataSourceConfig = require('./server/datasources.json');
+const db = new loopback.DataSource(dataSourceConfig[DATASOURCE_NAME]);
 ```
 
+Node.js does not allow the main module function to be async,
+the top-level code cannot use `await` keyword for flow control.
+As a workaround, we need to wrap the discovery code in a helper function
+that's executed using Promise API.
+
+```js
+discover().then(
+  success => process.exit(),
+  error => { console.error('UNHANDLED ERROR:\n', error); process.exit(1); },
```

Comment:
I'm not an expert on promises, but I thought errors were usually handled by `.catch()` chain. Is there a reason we prefer this style over the other?

Reply:
The difference between a two-args `then` and `catch` is in the case when a promise is resolved (successfully) and the first callback in `then` throws an error (e.g. `TypeError: undefined is not a function`). 

 - With two-arg `then`, this error correctly triggers `unhandledRejection`
 - With `.catch`, this error is passed to the `catch` callback, which has the effect that both "then" and "catch" callbacks are called.

Here is the situation where such behavior is not intended:

```ts
function legacyApiWithCallback(callback) {
  somePromiseApi()
    .then(() => callback())
    .catch(callback);
}
```

If `callback` throws an error, it will be called again to handle the error it has thrown.